seq/seq.1
.\"
.\" Copyright (c) 2023 Luiz Antônio Rangel
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.TH SEQ 1 "02/25/24" "Heirloom Toolchest" "User Commands"
.SH NAME
seq \- print a sequence of numbers
.SH SYNOPSIS
\fBseq\fR [\fB\-w\fR] [\fB\-ppicture\fR] [\fB\-sseparator\fR] [\fBstart\fR [\fBstep\fR]] stop
.SH DESCRIPTION
The
.B seq
utility prints a sequence of numbers, one per line
or separated per a character defined by the user, from
the \fBstart\fR (which is defaulted to 1) as near as
possible to \fBstop\fR, in steps of \fBstep\fR.
.br
It works in a similar spirit than a
.B for()
loop, present in the C, Go and
.B ksh93(1)
programming languages, and also
.B range()
in Python.
.PP
It accepts the following options:
.TP
.B \-w
Sets the number of decimal places as a constant width
per padding short numbers with leading decimal zeroes.
It gets annulled if combined with \-p.
.TP
.B \-p \fIpicture\fR
Defines the number of decimal places to be printed
after the decimal separator per using another number
as a \fIpicture\fR.
The default is full precision (see "%g" at \fBprintf(3)\fR).
.TP
.B \-s \fIseparator\fR
Defines a character to be the separator between the
numbers being printed. The default is a line break ("\\n").
.SH EXAMPLES
The
.B seq
utility may be used as one of the ways to calculate a
termial (\fIn?\fR) in a quick way using
.B bc(1)
in a
.B sh(1)
pipeline.
.IP \& 2
.BI "example% seq -s'+' 4 | bc"
.LP
.SH NOTES
.PP
It does not implement the \-f option found
in Plan 9 from Bell Labs' implementation,
although \-p can replace \-f in many cases.
.SH BUGS
Maybe?
.SH "SEE ALSO"
printf(3)
.SH HISTORY
A
.B seq
command first appeared in AT&T
Research Unix 8th edition.
.br
This version has its payloads
based off Research Unix 8th edition
with some extensions that appeared in
Plan 9 from Bell Labs, 1st Ed. and
GNU shell utilities 1.13 \(em for
the \-s option. The payload for
generating a reversed sequence
with an implicit decrement comes
from the
.B seq
implementation found on most BSDs
and Mac OS X, all of which were
derived from NetBSD 3.0.


CHANGES
Release ...
* watch: Use system(3) per default, with the previous exec(3) being moved to
  the "-x" option, also fix interval display at the information header;
* xargs: The new "-0" option now permits that a null character, '\0', is
  used as a delimiter instead of a new line;
* seq: Fix the step assumption so that if the stop is larger than the start,
  it will decrease one.

Release 250220
* seq, write: Fix memory fault due to illegal pointer access.
* cpio, pax: Fix manual page formatting.

Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."


seq/seq.c
/*
 * seq.c - print a sequence of numbers
 */
/*
 * Copyright (C) 2023-2025: Luiz Antônio Rangel (takusuman)
 *
 * SPDX-Licence-Identifier: Zlib
 */

#include <ctype.h>
#include <math.h>
#include <pfmt.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static char *progname;
/*
 * Make flags public since it will
 * be used by buildfmt().
 */
static bool fPicture = false,
	    fWadding = false;
static char *picstr = NULL;
static double start = 0.0,
	      stop = 0.0,
	      step = 0.0;

char *buildfmt(void);
char *getlgstr(void);
ssize_t afterdecsep(char s[]);
void usage(void);

int main(int argc, char *argv[]) {
	progname = argv[0];
	extern int optind;
	int option = 0;
	register double count = 0;
	char *format = NULL,
	     *separator = NULL;

	/*
	 * Stop searching for arguments in the moment
	 * it finds a digit, also stop increasing
	 * 'optind'.
	 */
	while (argv[optind] != NULL && !isdigit(argv[optind][1])
		&& (option = getopt(argc, argv, ":p:s:w")) != -1 ) {
		switch (option) {
			case 'p':
				fPicture = true;
				picstr = optarg;
				break;
			case 's':
				separator = optarg;
				break;
			case 'w':
				fWadding = true;
				break;
			default:
				usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc < 1)
		usage();

	/*
	 * If argc equals 1, stop will be defined as the
	 * first command line argument and start and step
	 * will be defaulted to 1.
	 * If argc is 2, stop will be defined as the
	 * second command line argument and start as the
	 * first. If the stop is smaller than the start,
	 * the step will be decreasing, not increasing.
	 * If argc is 3, stop will be defined as the
	 * third, start as the first and step as the
	 * second.
	 */
	stop = (argc == 1)
		? atof(argv[0])
		: (argc == 3)
			? atof(argv[2])
			: atof(argv[1]);
	start = (argc == 1)
		? 1
		: atof(argv[0]);
	step = (argc < 3)
		? (stop < start)
			? -1
			: 1
		: atof(argv[1]);

	if (step == 0) {
		pfmt(stderr, MM_ERROR,
			"%s: increment can not be zero.\n",
			progname);
		return 1;
	}

	format = buildfmt();

	/* If there's no separator set, defaults to a line break (\n). */
	separator = (! separator)
			? "\n"
			: separator;

	for (count = start; ((0 < step) ? (count <= stop) : (count >= stop));
							count += step) {
		/*
		 * If the count has come to the end or if the next sum
		 * is larger than stop, default separator back to '\n'.
		 * Do not be preoccupied about performance, the abs(3)
		 * family of functions is optimized at new processors
		 * and the overhead isn't so large anyway.
		 */
		separator = (fabs(count + step) > fabs(stop))
				? "\n" : separator;
		pfmt(stdout, (MM_NOSTD ^ MM_NOGET),
				format, count, separator);
	}
	free(format);

	return 0;
}

char *buildfmt(void) {
	char *picture = NULL,
	     *fmtbuf = NULL;

	if ((fmtbuf = calloc(32, sizeof(char))) == NULL) {
		pfmt(stderr, MM_ERROR,
			"%s: could not allocate an array of %d bytes.",
			progname, (32 * sizeof(char)));
		return NULL;
	}

	/* Default. */
	if (!fPicture && !fWadding) {
		snprintf(fmtbuf, sizeof(fmtbuf), "%s", "%g%s");
		return fmtbuf;
	}

	if (fPicture || fWadding) {
		ssize_t precision = 0;
		size_t natural = 0;

		char strnum[32] = "",
			/*
			 * Unlike the default,
			 * creating a buffer will
			 * be needed for avoiding
			 * truncating 'fmtbuf'.
			 */
		     buf[32] = "";

		/*
		 * Get how many digits after the
		 * decimal separator the picture
		 * have.
		 * If wadding is enabled alone, we
		 * shall check for the longest
		 * number on the buffer, if not,
		 * just the picture number.
		 */
		picture = (!fPicture && fWadding)
			? getlgstr()
			: picstr;
		precision = afterdecsep(picture);

		if (precision == -1) {
			pfmt(stderr, MM_ERROR,
				"%s: picture '%s' is not a number.\n",
				progname, picture);
			return NULL;
		}

		/*
		 * Basically the procedement
		 * that afterdecsep() does,
		 * but as a one-liner.
		 */
		if (fWadding) {
			snprintf(strnum, sizeof(strnum),
					"%.0f", ((start < stop || 0 < start)
								? stop
								: start));
		} else {
			snprintf(strnum, sizeof(strnum),
					"%.0f", atof(picture));
		}
		natural = strlen(strnum);

		/*
		 * free() only if picture comes
		 * from getlgstr(), since it had
		 * been allocated in memory, in
		 * contrast to picstr which is a
		 * vulgar string.
		 */
		if (!fPicture && fWadding) {
			free(picture);
		}

		/*
		 * If there are decimal values, add it to
		 * "natural" plus one, since "natural" in
		 * printf(3) context means the total
		 * number of characters that will be
		 * printed.
		 */
		if (precision > 0) {
			natural += precision;
			natural += 1;
		}

		/* Write to 'buf', then copy to fmtbuf. */
		snprintf(buf, sizeof(buf), "%%%d%ld.%ldf%%s",
					0, natural, precision);
		fmtbuf = strdup(buf);

		return fmtbuf;
	}

	return NULL;
}

char *getlgstr(void) {
	char strflt[32] = "",
	     *lgstnum = NULL;
	double fstn = 0.0,
	       stepn = 0.0;

	if ((lgstnum = calloc(sizeof(strflt), sizeof(char))) == NULL) {
		pfmt(stderr, MM_ERROR,
			"%s: could not allocate an array of %d bytes.",
			progname, (sizeof(strflt) * sizeof(char)));
		return NULL;
	}

	/*
	 * This method, although not being precise
	 * as the last, will get the most irregular
	 * number in the count per subtracting 'step'
	 * from the first number ('start'), which means
	 * getting the most of decimal numbers.
	 * It is better to subtract from the first
	 * number instead of the last because of numbers
	 * big enough to be converted by
	 * 'printf("%g", ...)' into scientific notation,
	 * which can not be parsed by afterdecsep().
	 * Also have a metric if the start isn't actually
	 * the stop (sequence done in decrescent order).
	 */
	fstn = (start < stop)
		? fabs(start)
		: fabs(stop);
	stepn = fabs(step);
	snprintf(strflt, sizeof(strflt), "%g", (fstn - stepn));
	lgstnum = strdup(strflt);

	return lgstnum;
}

ssize_t afterdecsep(char s[]) {
	size_t fraclen = 0;
	char *fracpart = NULL;

	if (isalpha(s[0])) {
		/* Panic. */
		return -1;
	}

	/*
	 * If there's not a decimal separator at *s,
	 * it's an invalid picture, but seq(1) will
	 * be not complaining about it.
	 */
	if ((fracpart = strchr(s, '.')) != NULL) {
		fracpart++;
		fraclen = strlen(fracpart);
	}

	return fraclen;
}

void usage(void) {
	pfmt(stderr, MM_NOSTD,
		"usage: %s: [-w] [-ppicture] [-sseparator] [start [step]] stop\n",
		progname);
	exit(1);
}
